File: decompiled/sources/com/hjq/widget/layout/NestedLinearLayout.java
package com.hjq.widget.layout;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.VelocityTracker;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.ViewParent;
import android.widget.LinearLayout;
import androidx.core.view.MotionEventCompat;
import androidx.core.view.NestedScrollingChild;
import androidx.core.view.NestedScrollingChildHelper;
import androidx.core.view.NestedScrollingParent;
import androidx.core.view.NestedScrollingParentHelper;

public class NestedLinearLayout extends LinearLayout implements NestedScrollingChild, NestedScrollingParent {
    private static final int INVALID_POINTER = -1;
    private int mActivePointerId;
    private final NestedScrollingChildHelper mChildHelper;
    private boolean mIsBeingDragged;
    private int mLastMotionY;
    private final float mMaximumVelocity;
    private final float mMinimumVelocity;
    private final NestedScrollingParentHelper mParentHelper;
    private final int[] mScrollConsumed;
    private final int[] mScrollOffset;
    private final float mTouchSlop;
    private VelocityTracker mVelocityTracker;

    public boolean onStartNestedScroll(View view, View view2, int i) {
        return (i & 2) != 0;
    }

    public NestedLinearLayout(Context context) {
        this(context, (AttributeSet) null, 0);
    }

    public NestedLinearLayout(Context context, AttributeSet attributeSet) {
        this(context, attributeSet, 0);
    }

    public NestedLinearLayout(Context context, AttributeSet attributeSet, int i) {
        super(context, attributeSet, i);
        this.mScrollConsumed = new int[2];
        this.mScrollOffset = new int[2];
        setWillNotDraw(false);
        this.mChildHelper = new NestedScrollingChildHelper(this);
        this.mParentHelper = new NestedScrollingParentHelper(this);
        setNestedScrollingEnabled(true);
        this.mTouchSlop = (float) ViewConfiguration.get(getContext()).getScaledTouchSlop();
        this.mMaximumVelocity = (float) ViewConfiguration.get(getContext()).getScaledMaximumFlingVelocity();
        this.mMinimumVelocity = (float) ViewConfiguration.get(getContext()).getScaledMinimumFlingVelocity();
    }

    public boolean onTouchEvent(MotionEvent motionEvent) {
        float f;
        MotionEvent obtain = MotionEvent.obtain(motionEvent);
        initVelocityTrackerIfNotExists();
        int actionMasked = obtain.getActionMasked();
        if (actionMasked != 0) {
            if (actionMasked != 1) {
                if (actionMasked == 2) {
                    int findPointerIndex = obtain.findPointerIndex(this.mActivePointerId);
                    if (findPointerIndex != -1) {
                        int y = (int) obtain.getY(findPointerIndex);
                        int i = this.mLastMotionY - y;
                        if (dispatchNestedPreScroll(0, i, this.mScrollConsumed, this.mScrollOffset)) {
                            i -= this.mScrollConsumed[1];
                            obtain.offsetLocation(0.0f, (float) this.mScrollOffset[1]);
                        }
                        if (!this.mIsBeingDragged && ((float) Math.abs(this.mLastMotionY - y)) > this.mTouchSlop) {
                            ViewParent parent = getParent();
                            if (parent != null) {
                                parent.requestDisallowInterceptTouchEvent(true);
                            }
                            this.mIsBeingDragged = true;
                            if (i > 0) {
                                f = ((float) i) - this.mTouchSlop;
                            } else {
                                f = ((float) i) + this.mTouchSlop;
                            }
                            i = (int) f;
                        }
                        int i2 = i;
                        if (this.mIsBeingDragged) {
                            this.mVelocityTracker.addMovement(motionEvent);
                            int[] iArr = this.mScrollOffset;
                            this.mLastMotionY = y - iArr[1];
                            if (dispatchNestedScroll(0, 0, 0, i2, iArr)) {
                                int i3 = this.mLastMotionY;
                                int[] iArr2 = this.mScrollOffset;
                                this.mLastMotionY = i3 - iArr2[1];
                                obtain.offsetLocation(0.0f, (float) iArr2[1]);
                            }
                        }
                    }
                } else if (actionMasked != 3) {
                    if (actionMasked == 5) {
                        int actionIndex = obtain.getActionIndex();
                        this.mLastMotionY = (int) obtain.getY(actionIndex);
                        this.mActivePointerId = obtain.getPointerId(actionIndex);
                    } else if (actionMasked == 6) {
                        onSecondaryPointerUp(obtain);
                        this.mLastMotionY = (int) obtain.getY(obtain.findPointerIndex(this.mActivePointerId));
                    }
                }
            }
            if (this.mIsBeingDragged) {
                this.mVelocityTracker.computeCurrentVelocity(1000, this.mMaximumVelocity);
                int yVelocity = (int) this.mVelocityTracker.getYVelocity(this.mActivePointerId);
                if (((float) Math.abs(yVelocity)) > this.mMinimumVelocity) {
                    flingWithNestedDispatch(-yVelocity);
                }
            }
            this.mActivePointerId = -1;
            endDrag();
        } else {
            this.mVelocityTracker.addMovement(motionEvent);
            this.mLastMotionY = (int) obtain.getY();
            this.mActivePointerId = obtain.getPointerId(0);
            startNestedScroll(2);
        }
        obtain.recycle();
        return true;
    }

    private void onSecondaryPointerUp(MotionEvent motionEvent) {
        int action = (motionEvent.getAction() & MotionEventCompat.ACTION_POINTER_INDEX_MASK) >> 8;
        if (motionEvent.getPointerId(action) == this.mActivePointerId) {
            int i = action == 0 ? 1 : 0;
            this.mLastMotionY = (int) motionEvent.getY(i);
            this.mActivePointerId = motionEvent.getPointerId(i);
            VelocityTracker velocityTracker = this.mVelocityTracker;
            if (velocityTracker != null) {
                velocityTracker.clear();
            }
        }
    }

    private void endDrag() {
        this.mIsBeingDragged = false;
        recycleVelocityTracker();
        stopNestedScroll();
    }

    private void flingWithNestedDispatch(int i) {
        float f = (float) i;
        if (!dispatchNestedPreFling(0.0f, f)) {
            dispatchNestedFling(0.0f, f, true);
        }
    }

    private void initVelocityTrackerIfNotExists() {
        if (this.mVelocityTracker == null) {
            this.mVelocityTracker = VelocityTracker.obtain();
        }
    }

    private void recycleVelocityTracker() {
        VelocityTracker velocityTracker = this.mVelocityTracker;
        if (velocityTracker != null) {
            velocityTracker.recycle();
            this.mVelocityTracker = null;
        }
    }

    public void requestDisallowInterceptTouchEvent(boolean z) {
        if (z) {
            recycleVelocityTracker();
        }
        super.requestDisallowInterceptTouchEvent(z);
    }

    public void setNestedScrollingEnabled(boolean z) {
        this.mChildHelper.setNestedScrollingEnabled(z);
    }

    public boolean isNestedScrollingEnabled() {
        return this.mChildHelper.isNestedScrollingEnabled();
    }

    public boolean startNestedScroll(int i) {
        return this.mChildHelper.startNestedScroll(i);
    }

    public void stopNestedScroll() {
        this.mChildHelper.stopNestedScroll();
    }

    public boolean hasNestedScrollingParent() {
        return this.mChildHelper.hasNestedScrollingParent();
    }

    public boolean dispatchNestedScroll(int i, int i2, int i3, int i4, int[] iArr) {
        return this.mChildHelper.dispatchNestedScroll(i, i2, i3, i4, iArr);
    }

    public boolean dispatchNestedPreScroll(int i, int i2, int[] iArr, int[] iArr2) {
        return this.mChildHelper.dispatchNestedPreScroll(i, i2, iArr, iArr2);
    }

    public boolean dispatchNestedFling(float f, float f2, boolean z) {
        return this.mChildHelper.dispatchNestedFling(f, f2, z);
    }

    public boolean dispatchNestedPreFling(float f, float f2) {
        return this.mChildHelper.dispatchNestedPreFling(f, f2);
    }

    public void onNestedScrollAccepted(View view, View view2, int i) {
        this.mParentHelper.onNestedScrollAccepted(view, view2, i);
        startNestedScroll(2);
    }

    public void onStopNestedScroll(View view) {
        this.mParentHelper.onStopNestedScroll(view);
        stopNestedScroll();
    }

    public void onNestedScroll(View view, int i, int i2, int i3, int i4) {
        dispatchNestedScroll(i, i2, i3, i4, (int[]) null);
    }

    public void onNestedPreScroll(View view, int i, int i2, int[] iArr) {
        dispatchNestedPreScroll(i, i2, iArr, (int[]) null);
    }

    public boolean onNestedFling(View view, float f, float f2, boolean z) {
        return dispatchNestedFling(f, f2, z);
    }

    public boolean onNestedPreFling(View view, float f, float f2) {
        return dispatchNestedPreFling(f, f2);
    }

    public int getNestedScrollAxes() {
        return this.mParentHelper.getNestedScrollAxes();
    }
}

File: decompiled/sources/com/hjq/widget/layout/NestedLinearLayout.kt
package com.hjq.widget.layout

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.VelocityTracker
import android.view.View
import android.view.ViewConfiguration
import android.view.ViewParent
import android.widget.LinearLayout
import androidx.core.view.NestedScrollingChild
import androidx.core.view.NestedScrollingChildHelper
import androidx.core.view.NestedScrollingParent
import androidx.core.view.NestedScrollingParentHelper
import kotlin.math.abs

open class NestedLinearLayout : LinearLayout, NestedScrollingChild, NestedScrollingParent {
    private var mActivePointerId = INVALID_POINTER
    private val mChildHelper: NestedScrollingChildHelper
    private var mIsBeingDragged = false
    private var mLastMotionY = 0
    private val mMaximumVelocity: Float
    private val mMinimumVelocity: Float
    private val mParentHelper: NestedScrollingParentHelper
    private val mScrollConsumed: IntArray
    private val mScrollOffset: IntArray
    private val mTouchSlop: Float
    private var mVelocityTracker: VelocityTracker? = null

    constructor(context: Context) : this(context, null, 0)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        mScrollConsumed = IntArray(2)
        mScrollOffset = IntArray(2)
        setWillNotDraw(false)
        mChildHelper = NestedScrollingChildHelper(this)
        mParentHelper = NestedScrollingParentHelper(this)
        isNestedScrollingEnabled = true
        val configuration = ViewConfiguration.get(context)
        mTouchSlop = configuration.scaledTouchSlop.toFloat()
        mMaximumVelocity = configuration.scaledMaximumFlingVelocity.toFloat()
        mMinimumVelocity = configuration.scaledMinimumFlingVelocity.toFloat()
    }

    override fun onStartNestedScroll(child: View, target: View, axes: Int): Boolean {
        return (axes and 2) != 0
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val obtain = MotionEvent.obtain(event)
        initVelocityTrackerIfNotExists()
        val actionMasked = obtain.actionMasked
        if (actionMasked != MotionEvent.ACTION_DOWN) {
            if (actionMasked != MotionEvent.ACTION_UP) {
                if (actionMasked == MotionEvent.ACTION_MOVE) {
                    val pointerIndex = obtain.findPointerIndex(mActivePointerId)
                    if (pointerIndex != -1) {
                        val y = obtain.getY(pointerIndex).toInt()
                        var deltaY = mLastMotionY - y
                        if (dispatchNestedPreScroll(0, deltaY, mScrollConsumed, mScrollOffset)) {
                            deltaY -= mScrollConsumed[1]
                            obtain.offsetLocation(0f, mScrollOffset[1].toFloat())
                        }
                        if (!mIsBeingDragged && abs(mLastMotionY - y).toFloat() > mTouchSlop) {
                            val parent: ViewParent? = parent
                            parent?.requestDisallowInterceptTouchEvent(true)
                            mIsBeingDragged = true
                            deltaY =
                                if (deltaY > 0) {
                                    (deltaY.toFloat() - mTouchSlop).toInt()
                                } else {
                                    (deltaY.toFloat() + mTouchSlop).toInt()
                                }
                        }
                        if (mIsBeingDragged) {
                            mVelocityTracker?.addMovement(event)
                            mLastMotionY = y - mScrollOffset[1]
                            if (dispatchNestedScroll(0, 0, 0, deltaY, mScrollOffset)) {
                                mLastMotionY -= mScrollOffset[1]
                                obtain.offsetLocation(0f, mScrollOffset[1].toFloat())
                            }
                        }
                    }
                } else if (actionMasked != MotionEvent.ACTION_CANCEL) {
                    if (actionMasked == MotionEvent.ACTION_POINTER_DOWN) {
                        val actionIndex = obtain.actionIndex
                        mLastMotionY = obtain.getY(actionIndex).toInt()
                        mActivePointerId = obtain.getPointerId(actionIndex)
                    } else if (actionMasked == MotionEvent.ACTION_POINTER_UP) {
                        onSecondaryPointerUp(obtain)
                        mLastMotionY = obtain.getY(obtain.findPointerIndex(mActivePointerId)).toInt()
                    }
                }
            }
            if (mIsBeingDragged) {
                mVelocityTracker?.computeCurrentVelocity(1000, mMaximumVelocity)
                val yVelocity = mVelocityTracker?.getYVelocity(mActivePointerId)?.toInt() ?: 0
                if (abs(yVelocity).toFloat() > mMinimumVelocity) {
                    flingWithNestedDispatch(-yVelocity)
                }
            }
            mActivePointerId = INVALID_POINTER
            endDrag()
        } else {
            mVelocityTracker?.addMovement(event)
            mLastMotionY = obtain.y.toInt()
            mActivePointerId = obtain.getPointerId(0)
            startNestedScroll(2)
        }
        obtain.recycle()
        return true
    }

    private fun onSecondaryPointerUp(event: MotionEvent) {
        val actionIndex = event.actionIndex
        if (event.getPointerId(actionIndex) == mActivePointerId) {
            val newIndex = if (actionIndex == 0) 1 else 0
            mLastMotionY = event.getY(newIndex).toInt()
            mActivePointerId = event.getPointerId(newIndex)
            mVelocityTracker?.clear()
        }
    }

    private fun endDrag() {
        mIsBeingDragged = false
        recycleVelocityTracker()
        stopNestedScroll()
    }

    private fun flingWithNestedDispatch(velocityY: Int) {
        val velocity = velocityY.toFloat()
        if (!dispatchNestedPreFling(0f, velocity)) {
            dispatchNestedFling(0f, velocity, true)
        }
    }

    private fun initVelocityTrackerIfNotExists() {
        if (mVelocityTracker == null) {
            mVelocityTracker = VelocityTracker.obtain()
        }
    }

    private fun recycleVelocityTracker() {
        val velocityTracker = mVelocityTracker
        if (velocityTracker != null) {
            velocityTracker.recycle()
            mVelocityTracker = null
        }
    }

    override fun requestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {
        if (disallowIntercept) {
            recycleVelocityTracker()
        }
        super.requestDisallowInterceptTouchEvent(disallowIntercept)
    }

    override fun setNestedScrollingEnabled(enabled: Boolean) {
        mChildHelper.isNestedScrollingEnabled = enabled
    }

    override fun isNestedScrollingEnabled(): Boolean {
        return mChildHelper.isNestedScrollingEnabled
    }

    override fun startNestedScroll(axes: Int): Boolean {
        return mChildHelper.startNestedScroll(axes)
    }

    override fun stopNestedScroll() {
        mChildHelper.stopNestedScroll()
    }

    override fun hasNestedScrollingParent(): Boolean {
        return mChildHelper.hasNestedScrollingParent()
    }

    override fun dispatchNestedScroll(
        dxConsumed: Int,
        dyConsumed: Int,
        dxUnconsumed: Int,
        dyUnconsumed: Int,
        offsetInWindow: IntArray?
    ): Boolean {
        return mChildHelper.dispatchNestedScroll(
            dxConsumed,
            dyConsumed,
            dxUnconsumed,
            dyUnconsumed,
            offsetInWindow
        )
    }

    override fun dispatchNestedPreScroll(
        dx: Int,
        dy: Int,
        consumed: IntArray?,
        offsetInWindow: IntArray?
    ): Boolean {
        return mChildHelper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow)
    }

    override fun dispatchNestedFling(velocityX: Float, velocityY: Float, consumed: Boolean): Boolean {
        return mChildHelper.dispatchNestedFling(velocityX, velocityY, consumed)
    }

    override fun dispatchNestedPreFling(velocityX: Float, velocityY: Float): Boolean {
        return mChildHelper.dispatchNestedPreFling(velocityX, velocityY)
    }

    override fun onNestedScrollAccepted(child: View, target: View, axes: Int) {
        mParentHelper.onNestedScrollAccepted(child, target, axes)
        startNestedScroll(2)
    }

    override fun onStopNestedScroll(target: View) {
        mParentHelper.onStopNestedScroll(target)
        stopNestedScroll()
    }

    override fun onNestedScroll(
        target: View,
        dxConsumed: Int,
        dyConsumed: Int,
        dxUnconsumed: Int,
        dyUnconsumed: Int
    ) {
        dispatchNestedScroll(dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed, null)
    }

    override fun onNestedPreScroll(target: View, dx: Int, dy: Int, consumed: IntArray) {
        dispatchNestedPreScroll(dx, dy, consumed, null)
    }

    override fun onNestedFling(
        target: View,
        velocityX: Float,
        velocityY: Float,
        consumed: Boolean
    ): Boolean {
        return dispatchNestedFling(velocityX, velocityY, consumed)
    }

    override fun onNestedPreFling(target: View, velocityX: Float, velocityY: Float): Boolean {
        return dispatchNestedPreFling(velocityX, velocityY)
    }

    override fun getNestedScrollAxes(): Int {
        return mParentHelper.nestedScrollAxes
    }

    companion object {
        private const val INVALID_POINTER = -1
    }
}
